components/greaselion/browser/greaselion_download_service.cc

```diff
@@ -185,21 +201,54 @@ GreaselionDownloadService::GreaselionDownloadService(
     is_dev_mode_ = true;
     resource_dir_ = forced_local_path;
     LoadDirectlyFromResourcePath();
+    dev_mode_task_runner_ = base::ThreadPool::CreateSequencedTaskRunner(
+        {base::TaskPriority::BEST_EFFORT,
+         base::TaskShutdownBehavior::CONTINUE_ON_SHUTDOWN, base::MayBlock()});
     dev_mode_path_watcher_ = std::make_unique<base::FilePathWatcher>();
-    if (!dev_mode_path_watcher_->Watch(resource_dir_,
-        true /*recursive*/,
-        base::Bind(&GreaselionDownloadService::OnDevModeLocalFileChanged,
-            weak_factory_.GetWeakPtr()))) {
-      LOG(ERROR) << "Greaselion could not watch filesystem for changes"
-          << " at path " << resource_dir_.LossyDisplayName();
-    }
+
+    using Callback = base::RepeatingCallback<void(bool)>;
+    base::FilePathWatcher::Callback file_path_watcher_callback =
+        base::BindRepeating(
+            [](scoped_refptr<base::SequencedTaskRunner> main_sequence,
+               const Callback& callback, const base::FilePath&, bool error) {
+              main_sequence->PostTask(FROM_HERE,
+                                      base::BindOnce(callback, error));
+            },
+            base::SequencedTaskRunnerHandle::Get(),
+            base::Bind(&GreaselionDownloadService::OnDevModeLocalFileChanged,
```

Comment:
Is there any way that we can avoid binding a callback into another callback?

No idea if this would work, but Is it perhaps possible to pass the `main_sequence` ref and a weak ptr to the `this` over to the task and then do something like this over on that task?

```cpp
main_sequence->PostTask(FROM_HERE,
    base::Bind(&GreaselionDownloadService::OnDevModeLocalFileChanged, etc...));
```

So that it's a little more clear where control is jumping to next?

Reply:
Yeah, I definitely see what you're saying. I tried a couple of runs at it this afternoon, but I can't seem to get a working combination that satisfies all of the necessary criteria (I either end up blocking or I end up on an invalid sequence).